File: src/ca/uqam/mgl7460/tp1/types/utils/Fabrique.java
package ca.uqam.mgl7460.tp1.types.utils;

import ca.uqam.mgl7460.tp1.implementations.utils.ConcreteFactory;
import ca.uqam.mgl7460.tp1.types.modeles.Adresse;
import ca.uqam.mgl7460.tp1.types.modeles.DemandePret;
import ca.uqam.mgl7460.tp1.types.modeles.DemandeurPret;
import ca.uqam.mgl7460.tp1.types.modeles.Propriete;
import ca.uqam.mgl7460.tp1.types.modeles.Resultat;
import ca.uqam.mgl7460.tp1.types.modeles.ResultatTraitement;
import ca.uqam.mgl7460.tp1.types.traitements.definitions.ConditionTransition;
import ca.uqam.mgl7460.tp1.types.traitements.definitions.DefinitionProcessus;
import ca.uqam.mgl7460.tp1.types.traitements.definitions.DefinitionTache;
import ca.uqam.mgl7460.tp1.types.traitements.definitions.DefinitionTransition;
import ca.uqam.mgl7460.tp1.types.traitements.definitions.TraitementTache;
import ca.uqam.mgl7460.tp1.types.traitements.instances.InstanceProcessus;
import ca.uqam.mgl7460.tp1.types.traitements.instances.InstanceTache;

public interface Fabrique {

    /**
     * Créer une propriété, en fournissant son adresse comme argument
     * @param adresse
     * @return
     */
    public Propriete creerPropriete(Adresse adresse);

    /*
     * creer une propriété en fournissant son adresse et sa valeur de marché
     */
    public Propriete creerPropriete(Adresse adresse, float valeurMarche);

    /**
     * Créer un-e demandeur-se de prêt en fournissant prenom, nom, et numéro d'assurance sociale (NAS)
     * @param prenom
     * @param nom
     * @param NAS
     * @return
     */
    public DemandeurPret creerDemandeurPret(String prenom, String nom, String NAS);

    /**
     * Créer un-e demandeur-se de prêt en founrissant prenom, nom, NAS, revenu annuel, obligations annuelles, et score de crédit
     * @param prenom
     * @param nom
     * @param NAS
     * @param revenuAnnuel
     * @param obligationsAnnuelles
     * @param scoreCredit
     * @return
     */
    public DemandeurPret creerDemandeurPret(String prenom, String nom, String NAS, float revenuAnnuel, float obligationsAnnuelles, int scoreCredit);

    /**
     * Créer une demande de prêt en fournissant, en argument, la propriété et le-la demandeur-se
     * @param propriete
     * @param demandeurPret
     * @return
     */
    public DemandePret creerDemandePret(Propriete propriete, DemandeurPret demandeurPret);

    /**
     * Créer une demande de prêt en fournissant la propriété, le-la demandeur-se de prêt, le prix d'achat, et 
     * la mise de fonds initiale de l'emprunteur-se.
     * @param propriete
     * @param demandeurPret
     * @param prixAchat
     * @param miseDeFonds
     * @return
     */
    public DemandePret creerDemandePret(Propriete propriete, DemandeurPret demandeurPret, float prixAchat, float miseDeFonds);

    /**
     * Créer un resultat de traitement en fournissant le resultat (une valeur
     * parmi NONDETERMINE, ACCEPTEE, REFUSEE)
     * @param resultat
     * @return
     */
    public ResultatTraitement creerResultatTraitement(Resultat resultat);

    /**
     * Créer une définition de tache, en fournissant le nom et la description de la tache
     * @param nom
     * @param description
     * @return
     */
    public DefinitionTache creerDefinitionTache(String nom, String description);

    /**
     * Créer une définition de tache, en fournissant le nom, la description de la tache, ET
     * la fonction à exécuter par la tâche (qui est représentée par une variable du type
     * TraitementTache qui est une 'functional interface')
     * @param nom
     * @param description
     * @param traitementTache
     * @return
     */
    public DefinitionTache creerDefinitionTache(String nom, String description, TraitementTache traitementTache);

    /**
     * Créer une définition (spécification) de transition entre deux (définitions de) tâche, en
     * fournissant: 1) la tacheSource, 2) la tacheDestination, et 3) la conditionTransition, qui est
     * une 'interface fonctionnelle'
     * @param tacheSource
     * @param tacheDestination
     * @param conditionTransition
     * @return
     */
    public DefinitionTransition creerDefinitionTransition(DefinitionTache tacheSource, DefinitionTache tacheDestination,
            ConditionTransition conditionTransition);

    /**
     * Créer une définition (spécification) de transition en entre deux définitions de tâches, en fournissant la tacheSource
     * et la sourceDestination. Ici, cela suppose que la condition de transition est toujours vraie
     * @param tacheSource
     * @param tacheDestination
     * @return
     */
    public DefinitionTransition creerDefinitionTransition(DefinitionTache tacheSource, DefinitionTache tacheDestination);

    /**
     * Crèer une coquille de définition de processus, en fournissant le nom et la description du processus
     * @param nomProcessus
     * @param descriptionProcessus
     * @return
     */
    public DefinitionProcessus creerDefinitionProcessus(String nomProcessus, String descriptionProcessus);

    /**
     * Créer une instance de processus (que l'on peut donc exécuter par la suite) en fournissant, a) la
     * definitionProcess à exécuter, et b) la demandePret sur laquelle exécuter le processus.
     * @param definitionProcessus
     * @param demandePret
     * @return
     */
    public InstanceProcessus creerInstanceProcessus(DefinitionProcessus definitionProcessus, DemandePret demandePret);

    /**
     * Créer une instance de tâche en fournissant l'instanceProcessus dont elle va faire partie, et la 
     * definitionTache dont elle sera une copie/instance
     * @param instanceProcessus
     * @param definitionTache
     * @return
     */
    public InstanceTache creerInstanceTache(InstanceProcessus instanceProcessus, DefinitionTache definitionTache);

    /**
     * Cette méthode retourne un singleton de la Fabrique. C'est un "trick" pas trop catholique (méthode statique concrête
     * dans un interface), mais ça permet aux classes d'implantation de faire appel au singleton de la factory sans
     * "connaitre" la classe d'implementation de la fabrique.
     * 
     * Quand vous aurez implémenté la classe d'implémentation de la Fabrique, vous mettrez son nom à la place de 
     * <code>FabriqueImpl</code>, qui est le nom que j'ai choisi.
     * @return
     */
    public static Fabrique getSingletonFabrique() {
        Fabrique singleton = null;
        singleton = ConcreteFactory.getSingleton();
        return singleton;
    }
}

File: src_python/utils/fabrique.py

from modeles.adresse import Adresse
from modeles.demande_pret import DemandePret
from modeles.demandeur_pret import DemandeurPret
from modeles.propriete import Propriete
from modeles.resultat import Resultat
from modeles.resultat_traitement import ResultatTraitement
from traitements.definitions.condition_transition import ConditionTransition
from traitements.definitions.definition_processus import DefinitionProcessus
from traitements.definitions.definition_tache import DefinitionTache
from traitements.definitions.definition_transition import DefinitionTransition
from traitements.definitions.traitement_tache import TraitementTache
from traitements.instances.instance_processus import InstanceProcessus
from traitements.instances.instance_tache import InstanceTache


def creer_propriete(adresse: Adresse, valeur_marche: float = 0.0) -> Propriete:
    return Propriete(adresse, valeur_marche)

def creer_demandeur_pret(prenom: str, nom: str, nas: str,
                         revenu_annuel: float | None = None,
                         obligations_annuelles: float | None = None,
                         score_credit: int | None = None) -> DemandeurPret:
    return DemandeurPret(
        prenom,
        nom,
        nas,
        revenu_annuel or 0.0,
        obligations_annuelles or 0.0,
        score_credit or 0
    )

def creer_demande_pret(propriete: Propriete, demandeur_pret: DemandeurPret, prix_achat=0.0, montant_mise_de_fonds=0.0) -> DemandePret:
    return DemandePret(propriete=propriete, demandeur_pret=demandeur_pret, prix_achat=prix_achat, montant_mise_de_fonds=montant_mise_de_fonds)

def creer_resultat_traitement(resultat: Resultat) -> ResultatTraitement:
    return ResultatTraitement(resultat)

def creer_definition_tache(nom: str, description: str, traitementTache: TraitementTache = None) -> DefinitionTache:
    return DefinitionTache(nom=nom, description=description, traitementTache=traitementTache)


def creer_definition_transition(tache_source, tache_destination, condition_transition: ConditionTransition = None) -> DefinitionTransition:
    return DefinitionTransition(tache_source = tache_source, tache_destination = tache_destination, condition_transition = condition_transition)

def creer_definition_processus(nom: str, description: str) -> DefinitionProcessus:
    return DefinitionProcessus(nom=nom, description=description)

def creer_instance_processus(definition_processus: DefinitionProcessus, demandePret: DemandePret) -> InstanceProcessus:
    return InstanceProcessus(definition_processus = definition_processus, demandePret = demandePret)

def creer_instance_tache(processus_englobant: InstanceProcessus, definition_tache: DefinitionTache) -> InstanceTache:
    return InstanceTache(definition_tache=definition_tache, processus_englobant=processus_englobant)